Add Trace, update for new order of analysis (?)

rmosolgo opened the pull request:
This is to try out a new tracing API (https://github.com/rmosolgo/graphql-ruby/pull/4344). It looks like it will work fine.

Along the way, I found that some tests here fail on `master`, presumably because my refactors to Language::Visitor and Analysis::AST::Visitor (https://github.com/rmosolgo/graphql-ruby/pull/4338) changed the order of these arrays. Is that OK?

rmosolgo commented:
(I signed the CLA, but I don't see an option to re-run the job myself.)

rmosolgo commented on lib/graphql/metrics/trace.rb:
```diff
@@ -0,0 +1,190 @@
+      def initialize(**_rest)
+        super
+        context = (@multiplex || @query).context
+        @skip_tracing = context&.fetch(GraphQL::Metrics::SKIP_GRAPHQL_METRICS_ANALYSIS, false)
```
I'm not sure how to implement it... 

For performance's sake, each tracer is a module that's mixed into a class. Flow passes from tracer to tracer via `super` calls. (This is like `Language::Visitor`.) But it means there's no easy way to exclude one module's code from the resulting code flow 🤔

swalkinshaw replied:
Ah yeah I can't see a way around that. Not a big deal anyway I assume this use case is pretty specific to a library

rmosolgo replied:
Oh -- there _is_ a way to do it. The application could pass in `context[:trace]` to use instead of `schema.trace_class.new`, allowing the application to customize the trace per-request. Maybe you could dress that up a bit. 🤔

rmosolgo commented:
:+1: I updated it to sort those arrays before comparing them, so it should pass on all versions. And I added a new gemfile to run tests on my WIP branch.

swalkinshaw commented:
🎉  looks good to me. The method based tracer is a much better interface

rmosolgo commented:
Thanks for taking a look. I added a benchmark result above. I'll sleep on it and merge that graphql-ruby PR soon, then update this accordingly :+1:

swalkinshaw commented:
@rmosolgo this is good to go after a rebase right?

rmosolgo commented on lib/graphql/metrics/tracer.rb:
```diff
@@ -54,11 +60,13 @@ def trace(key, data, &block)
         end
       end
+<<<<<<< HEAD
+=======
```
derp, thanks!

rmosolgo commented:
😅 Alright, tests pass locally!

toneymathews reviewed (approved):
looks good to me.
